doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -848,37 +1228,427 @@ The Ltac2 parser can be extended with syntactic notations.
+example that passes two parameters to a missing build-in tactic.
+
+.. _syntactic_classes:
+
+Syntactic classes
+~~~~~~~~~~~~~~~~~
+
+The simplest syntactic classes in Ltac2 notations represent individual nonterminals
+from the |Coq| grammar.  Only a few selected nonterminals are available as syntactic classes.
+In addition, there are metasyntactic operations for describing
+more complex syntax, such as making an item optional or representing a list of items.
+When parsing, each syntactic class expression returns a value that's bound to a name in the
+notation definition.
+
+Syntactic classes are described with a form of S-expression:
+
+   .. insertprodn ltac2_scope ltac2_scope
+
+   .. prodn::
+      ltac2_scope ::= @string
+      | @int
+      | @name
+      | @name ( {+, @ltac2_scope } )
+
+.. todo no syn class for ints or strings?
+   parm names are not reserved (e.g the var can be named "list1")
+
+Metasyntactic operations that can be applied to other syntactic classes are:
+
+  :n:`opt(@ltac2_scope)`
+    Parses an optional :token:`ltac2_scope`.  The associated value is either :n:`None` or
+    enclosed in :n:`Some`
+
+  :n:`list1(@ltac2_scope {? , @string })`
+    Parses a list of one or more :token:`ltac2_scope`\s.  If :token:`string` is specified,
+    items must be separated by :token:`string`.
+
+  :n:`list0(@ltac2_scope {? , @string })`
+    Parses a list of zero or more :token:`ltac2_scope`\s.  If :token:`string` is specified,
+    items must be separated by :token:`string`.  For zero items, the associated value
+    is an empty list.
+
+  :n:`seq({+, @ltac2_scope })`
+    Parses the :token:`ltac2_scope`\s in order.  The associated value is a tuple,
+    omitting :token:`ltac2_scope`\s that are :token:`string`\s.
+    `self` and `next` are not permitted within `seq`.
+
+The following classes represent nonterminals with some special handling.  The
+table further down lists the classes that that are handled plainly.
+
+  :n:`constr {? ( {+, @scope_key } ) }`
+    Parses a :token:`term`.  If specified, the :token:`scope_key`\s are used to interpret
+    the term (as described in  :ref:`LocalInterpretationRulesForNotations`).  The last
+    :token:`scope_key` is the top of the scope stack that's applied to the :token:`term`.
+
+    .. todo PR the original text describes "constr:(c%A%B)" which I couldn't make sense of
+       Is that something we should keep (and if so, what is an example?)
+
+  :n:`open_constr`
+    Parses an open :token:`term`.
+
+  :n:`ident`
+    Parses :token:`ident` or :n:`$@ident`.  The first form returns :n:`ident:(@ident)`,
+    while the latter form returns the variable :n:`@ident`.
+
+  :n:`tactic {? (@int) }`
+    Parses an :token:`ltac2_expr`.  If :token:`int` is specified, the construct
+    parses a :n:`ltac2_expr@int`, for example `tactic(5)` parses :token:`ltac2_expr5`.
+    `tactic(6)` parses :token:`ltac2_expr`.
+    :token:`int` must be in the range `0 .. 6`.
```

Comment:
```suggestion
    Parses an :token:`ltac2_expr`.  If :token:`int` is specified, the construct
    parses a :n:`ltac2_expr@int`, for example `tactic(5)` parses :token:`ltac2_expr5`.
    `tactic(6)` parses :token:`ltac2_expr`.
    :token:`int` must be in the range `0 .. 6`.
    
    Note that the syntax for integers, strings, tuples, and lists is part of :token:`ltac2_expr`,
    so you can use `tactic` scope to parse these constructs.
```

This will fix #12805